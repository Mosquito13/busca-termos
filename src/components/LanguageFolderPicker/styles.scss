@import '../../sass/styles.scss';

.language-folder-picker {
    display: flex;
    flex-direction: column;

    &__fields {
        display: flex;
    }

    &__error {
        display: flex;
        align-items: center;
        color: $color-danger-dark;
        font-size: 1.2rem;
        padding: .5rem .5rem 0;
        height: 1.5rem;
        opacity: 1;
        transition: height .2s ease-out, opacity .5s;

        &--hidden {
            height: 0;
            opacity: 0;
        }
    }

    &__error-message {
        margin-left: .5rem;
    }

    &__button {
        margin-left: 1rem;
        display: flex;
    }

    &__input-file {
        display: none;
    }
}