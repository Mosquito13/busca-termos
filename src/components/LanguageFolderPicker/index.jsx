import { useCallback, useRef } from 'react';
import { FiXOctagon, FiMoreHorizontal } from 'react-icons/fi';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import Icon, { EXTRA_SMALL } from '../common/Icon';
import InputText from '../common/InputText';
import Button from '../common/Button';

import './styles.scss';

const getErrorMessage = (error) => {
  if (!Boolean(error)) {
    return '';
  }

  return (
    <>
      <div className="language-folder-picker__error-icon">
        <Icon size={EXTRA_SMALL}>
          <FiXOctagon />
        </Icon>
      </div>
      <div className="language-folder-picker__error-message">{error}</div>
    </>
  );
};

const LanguageFolderPicker = ({ value, error, onChange, disabled }) => {
  const inputFileRef = useRef();

  const handleClickButtonFind = useCallback(() => {
    inputFileRef.current.click();
  }, []);

  const onChangeFolder = useCallback(
    (event) => {
      if (event.currentTarget.files.length === 0) {
        return;
      }

      const folderPath = event.currentTarget.files[0].path.split('\\');

      folderPath.pop();
      onChange(folderPath.join('/'));
    },
    [onChange]
  );

  const errorClasses = classNames(
    'language-folder-picker__error',
    !Boolean(error) && 'language-folder-picker__error--hidden'
  );

  return (
    <div className="language-folder-picker">
      <div className="language-folder-picker__fields">
        <InputText
          value={value}
          onChange={onChange}
          placeholder="Caminho atÃ© a pasta language..."
          disabled={disabled}
        />
        <div className="language-folder-picker__button">
          <Button
            icon={
              <Icon>
                <FiMoreHorizontal />
              </Icon>
            }
            tooltip="Selecionar pasta..."
            onClick={handleClickButtonFind}
            disabled={disabled}
          />
        </div>
        <input
          className="language-folder-picker__input-file"
          onChange={onChangeFolder}
          ref={inputFileRef}
          type="file"
          id="filepicker"
          name="fileList"
          webkitdirectory=""
          directory=""
        />
      </div>
      <div className={errorClasses}>{getErrorMessage(error)}</div>
    </div>
  );
};

LanguageFolderPicker.propTypes = {
  /**
   * Valor do campo
   */
  value: PropTypes.string,
  /**
   * Mensagem de erro no campo
   */
  error: PropTypes.string,
  /**
   * Funcao para trocar o valor do campo
   */
  onChange: PropTypes.func,
  /**
   * Indica se esta desabilitado
   */
  disabled: PropTypes.bool
};

export default LanguageFolderPicker;
