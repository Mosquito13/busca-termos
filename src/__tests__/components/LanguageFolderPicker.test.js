import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import * as reactRedux from 'react-redux';

import LanguageFolderPicker from '../../components/LanguageFolderPicker';

describe(`${LanguageFolderPicker.name}`, () => {
  describe('tests', () => {
    let wrapper, onChangeSpy;

    beforeAll(() => {
      onChangeSpy = jest.fn();

      wrapper = shallow(<LanguageFolderPicker onChange={onChangeSpy} />);
    });

    it('should render', () => {
      expect(wrapper.isEmptyRender()).toBeFalsy();
      expect(wrapper.find('.language-folder-picker')).toHaveLength(1);
      expect(
        wrapper.find('.language-folder-picker__error-message')
      ).toHaveLength(0);
      expect(wrapper.find('[data-testid="lfp-input"]')).toHaveLength(1);
      expect(wrapper.find('[data-testid="lfp-btn"]')).toHaveLength(1);
    });

    it('should not fire onChange when nothing was selected on the input', () => {
      wrapper.find('.language-folder-picker__input-file').simulate('change', {
        currentTarget: {
          files: []
        }
      });
      expect(onChangeSpy).toBeCalledTimes(0);
    });

    it('should fire onChange with the path', () => {
      wrapper.find('.language-folder-picker__input-file').simulate('change', {
        currentTarget: {
          files: [
            {
              path: 'C:\\aFolder\\anotherFolder\\fileName.txt'
            }
          ]
        }
      });
      expect(onChangeSpy).toBeCalledTimes(1);
      expect(onChangeSpy).toBeCalledWith('C:/aFolder/anotherFolder');
    });

    it('should render error message', () => {
      wrapper.setProps({ error: 'An error message ' });
      expect(
        wrapper.find('.language-folder-picker__error-message')
      ).toHaveLength(1);
      expect(wrapper.text()).toContain('An error message');
    });

    afterAll(() => {
      wrapper.unmount();
    });
  });

  describe('snapshots', () => {
    beforeAll(() => {
      jest.spyOn(reactRedux, 'useSelector').mockImplementation(() => false);
    });

    it('should render without value', () => {
      const tree = renderer.create(<LanguageFolderPicker />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render with value', () => {
      const tree = renderer
        .create(<LanguageFolderPicker value="C:/folder/anotherFolder/" />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render with error', () => {
      const tree = renderer
        .create(<LanguageFolderPicker error="An error" />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render disabled', () => {
      const tree = renderer.create(<LanguageFolderPicker disabled />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
