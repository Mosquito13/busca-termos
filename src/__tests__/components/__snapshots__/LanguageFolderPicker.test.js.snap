// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageFolderPicker snapshots should render disabled 1`] = `
<div
  className="language-folder-picker"
>
  <div
    className="language-folder-picker__fields"
  >
    <div
      className="input-text"
    >
      <input
        className="input-text__input input-text__input--disabled"
        disabled={true}
        onChange={[Function]}
        placeholder="Caminho até a pasta language..."
        spellCheck={false}
        type="text"
        value=""
      />
    </div>
    <div
      className="language-folder-picker__button"
    >
      <button
        className="button button--with-icon button--disabled"
        disabled={true}
        onClick={[Function]}
        title="Selecionar pasta..."
      >
        <div
          className="button__content"
        >
          <div
            className="icon icon--medium"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="1"
              />
              <circle
                cx="19"
                cy="12"
                r="1"
              />
              <circle
                cx="5"
                cy="12"
                r="1"
              />
            </svg>
          </div>
        </div>
      </button>
    </div>
    <input
      className="language-folder-picker__input-file"
      directory=""
      id="filepicker"
      name="fileList"
      onChange={[Function]}
      type="file"
      webkitdirectory=""
    />
  </div>
  <div
    className="language-folder-picker__error language-folder-picker__error--hidden"
  >
    
  </div>
</div>
`;

exports[`LanguageFolderPicker snapshots should render with error 1`] = `
<div
  className="language-folder-picker"
>
  <div
    className="language-folder-picker__fields"
  >
    <div
      className="input-text"
    >
      <input
        className="input-text__input"
        onChange={[Function]}
        placeholder="Caminho até a pasta language..."
        spellCheck={false}
        type="text"
        value=""
      />
    </div>
    <div
      className="language-folder-picker__button"
    >
      <button
        className="button button--with-icon"
        onClick={[Function]}
        title="Selecionar pasta..."
      >
        <div
          className="button__content"
        >
          <div
            className="icon icon--medium"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="1"
              />
              <circle
                cx="19"
                cy="12"
                r="1"
              />
              <circle
                cx="5"
                cy="12"
                r="1"
              />
            </svg>
          </div>
        </div>
      </button>
    </div>
    <input
      className="language-folder-picker__input-file"
      directory=""
      id="filepicker"
      name="fileList"
      onChange={[Function]}
      type="file"
      webkitdirectory=""
    />
  </div>
  <div
    className="language-folder-picker__error"
  >
    <div
      className="language-folder-picker__error-icon"
    >
      <div
        className="icon icon--extra-small"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polygon
            points="7.86 2 16.14 2 22 7.86 22 16.14 16.14 22 7.86 22 2 16.14 2 7.86 7.86 2"
          />
          <line
            x1="15"
            x2="9"
            y1="9"
            y2="15"
          />
          <line
            x1="9"
            x2="15"
            y1="9"
            y2="15"
          />
        </svg>
      </div>
    </div>
    <div
      className="language-folder-picker__error-message"
    >
      An error
    </div>
  </div>
</div>
`;

exports[`LanguageFolderPicker snapshots should render with value 1`] = `
<div
  className="language-folder-picker"
>
  <div
    className="language-folder-picker__fields"
  >
    <div
      className="input-text"
    >
      <input
        className="input-text__input"
        onChange={[Function]}
        placeholder="Caminho até a pasta language..."
        spellCheck={false}
        type="text"
        value="C:/folder/anotherFolder/"
      />
    </div>
    <div
      className="language-folder-picker__button"
    >
      <button
        className="button button--with-icon"
        onClick={[Function]}
        title="Selecionar pasta..."
      >
        <div
          className="button__content"
        >
          <div
            className="icon icon--medium"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="1"
              />
              <circle
                cx="19"
                cy="12"
                r="1"
              />
              <circle
                cx="5"
                cy="12"
                r="1"
              />
            </svg>
          </div>
        </div>
      </button>
    </div>
    <input
      className="language-folder-picker__input-file"
      directory=""
      id="filepicker"
      name="fileList"
      onChange={[Function]}
      type="file"
      webkitdirectory=""
    />
  </div>
  <div
    className="language-folder-picker__error language-folder-picker__error--hidden"
  >
    
  </div>
</div>
`;

exports[`LanguageFolderPicker snapshots should render without value 1`] = `
<div
  className="language-folder-picker"
>
  <div
    className="language-folder-picker__fields"
  >
    <div
      className="input-text"
    >
      <input
        className="input-text__input"
        onChange={[Function]}
        placeholder="Caminho até a pasta language..."
        spellCheck={false}
        type="text"
        value=""
      />
    </div>
    <div
      className="language-folder-picker__button"
    >
      <button
        className="button button--with-icon"
        onClick={[Function]}
        title="Selecionar pasta..."
      >
        <div
          className="button__content"
        >
          <div
            className="icon icon--medium"
          >
            <svg
              fill="none"
              height="1em"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="1"
              />
              <circle
                cx="19"
                cy="12"
                r="1"
              />
              <circle
                cx="5"
                cy="12"
                r="1"
              />
            </svg>
          </div>
        </div>
      </button>
    </div>
    <input
      className="language-folder-picker__input-file"
      directory=""
      id="filepicker"
      name="fileList"
      onChange={[Function]}
      type="file"
      webkitdirectory=""
    />
  </div>
  <div
    className="language-folder-picker__error language-folder-picker__error--hidden"
  >
    
  </div>
</div>
`;
